File: src/199-binarytree-right-side-view/java/BinaryTreeRightSideView.java
import java.util.ArrayList;
import java.util.List;

public class BinaryTreeRightSideView {
    // intuition answer via recrursion
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        viewRight(root, ans, 0);
        return ans;
    }

    public void viewRight(TreeNode node, List<Integer> ans, int currDepth) {
        if (node == null)
            return;
        if (currDepth == ans.size())
            ans.add(node.val);

        viewRight(node.right, ans, currDepth + 1);
        viewRight(node.left, ans, currDepth + 1);

    }

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: src/199-binarytree-right-side-view/kotlin/BinaryTreeRightSideView.kt
class BinaryTreeRightSideView {
    //solution based on Java Solution
    fun rightSideView(root: TreeNode?): List<Int> {
        val ans = arrayListOf<Int>()
        viewRight(root, ans, 0)
        return ans
    }

    fun viewRight(node: TreeNode?, ans: MutableList<Int>, currDepth: Int) {
        if (node == null)
            return
        if (currDepth == ans.size)
            ans.add(node?.`val` ?: 0)
    
        viewRight(node.right, ans, currDepth + 1)
        viewRight(node.left, ans, currDepth + 1)
    }

    //one liner kotlin code
    fun rightSideView(root: TreeNode?): List<Int> = root ?.
        let{ listOf(it.`val`) + rightSideView(it.right).
            let { it + rightSideView(root.left).drop(it.size) } } ?: emptyList()
}

class TreeNode(var `val`: Int) {
      var left: TreeNode? = null
      var right: TreeNode? = null
}